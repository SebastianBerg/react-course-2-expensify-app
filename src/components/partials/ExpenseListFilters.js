import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from "../../actions/filters";
import { Link } from "react-router-dom";

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calenderFocused => {
    this.setState(() => ({ calenderFocused }));
  };

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    } else {
      return null;
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="header__content--s-padding">
          <div className="page-header__actions">
            <Link className="button show-for-mobile" to="/create-post">
              Add Expense
            </Link>
          </div>
          <h1 className="page-header__title">Filter expenses to view</h1>
          <div className="page-header__actions">
            <Link className="button show-for-desktop" to="/create-post">
              Add Expense
            </Link>
          </div>
        </div>

        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              value={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder="Search expenses"
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              // value={this.props.filters.sortBy}
              onChange={this.onSortChange}
              defaultValue={"default"}
            >
              <option value="default" disabled>
                Sort By
              </option>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calenderFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
              startDateId={"123456"}
              endDateId={"654321"}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
