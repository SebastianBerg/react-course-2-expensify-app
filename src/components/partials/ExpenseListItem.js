import React from "react";
import { Link } from "react-router-dom";

export function ExpenseListItem({ id, description, amount, createdAt }) {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>
        {" "}
        Amount: {amount} Created At: {createdAt}
      </p>
    </div>
  );
}

export default ExpenseListItem;
