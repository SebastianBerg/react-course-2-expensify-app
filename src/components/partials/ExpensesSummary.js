import React from "react";
import { connect } from "react-redux";
import getVisibleExpenses from "../../selectors/expenses";
import selectExpensesTotal from "../../selectors/expenses-total";
import numeral from "numeral";

export class ExpensesSummary extends React.Component {
  getExpensesNumber() {
    const projectArray = Array.from(this.props.expenses);
    if (projectArray) {
      const expensesNumber = projectArray.length;
      return expensesNumber;
    }
  }

  render() {
    const { expenseCount, expensesTotal } = this.props;

    let expenseWord = expenseCount === 1 ? "expense" : "expenses";

    return (
      <div className="page-header">
        <div className="content-container">
          <div>
            {this.props.expenses.length > 0 ? (
              <h1 className="page-header__title">
                Showing <span>{expenseCount}</span>of
                <span> {this.getExpensesNumber()}</span>
                {expenseWord} for{" "}
                <span>{numeral(expensesTotal / 100).format("$0,0")}</span>
              </h1>
            ) : null}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses),
    expenses: state.expenses
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
