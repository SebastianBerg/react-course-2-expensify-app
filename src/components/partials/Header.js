import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { startLogout } from "../../actions/auth";

export class Header extends Component {
  render() {
    const { startLogout } = this.props;

    return (
      <header className="header">
        <div className="content-container">
          <div className="header__content">
            <div>
              <Link className="header__title" to="/dashboard">
                <h1>Expensify</h1>
              </Link>
            </div>
            <div>
              {this.props.location.pathname !== "/dashboard" ? (
                <Link to="/dashboard" className="button button--link">
                  Home
                </Link>
              ) : (
                <Link to="/create-project" className="button button--link">
                  Create New Project
                </Link>
              )}

              <button className="button button--link" onClick={startLogout}>
                Logout
              </button>
            </div>
          </div>
        </div>
      </header>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout())
});

export default withRouter(
  connect(
    undefined,
    mapDispatchToProps
  )(Header)
);
