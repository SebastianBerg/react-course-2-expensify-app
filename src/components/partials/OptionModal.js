import React from "react";
import ReactModal from "react-modal";

const OptionModal = props => (
  <ReactModal
    isOpen={!!props.isOpen}
    ariaHideApp={false}
    onRequestClose={props.handleCloseModal}
  >
    <div className="modal-box">
      <h1 className="modal-title">
        Are you sure you want to remove this {props.removalObject}?
      </h1>
      <div className="modal-buttonbox">
        <button onClick={props.onRemove} className="button">
          Yes, I'm sure.
        </button>
        <button onClick={props.handleCloseModal} className="button--light-grey">
          No, I want to keep the {props.removalObject}.
        </button>
      </div>
    </div>
  </ReactModal>
);

export default OptionModal;
