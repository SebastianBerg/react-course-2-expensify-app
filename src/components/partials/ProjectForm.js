import React, { Component } from "react";
import "react-dates/initialize";

class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.project ? props.project.name : "",
      initialFunds: props.project ? props.project.initialFunds : "",
      info: props.project ? props.project.info : "",
      calenderFocused: false,
      error: ""
    };
  }

  onDescriptionChange = e => {
    const name = e.target.value;
    this.setState(() => ({ name }));
  };

  onAmountChange = e => {
    const initialFunds = e.target.value;

    if (!initialFunds || initialFunds.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ initialFunds }));
    }
  };

  onInfoChange = e => {
    const info = e.target.value;
    this.setState(() => ({ info }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.name || !this.state.initialFunds) {
      this.setState(() => ({
        error: "Please provide a Name and an initial Funding. (This can be 0)"
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        name: this.state.name,
        initialFunds: this.state.initialFunds,
        info: this.state.info
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          className="text-input"
          name=""
          id=""
          placeholder="Name"
          autoFocus
          value={this.state.name}
          onChange={this.onDescriptionChange}
        />

        <input
          type="amount"
          className="text-input"
          placeholder="Initial Funds"
          value={this.state.initialFunds}
          onChange={this.onAmountChange}
        />

        <textarea
          placeholder="Add info about the project (optional)"
          className="textarea"
          cols="30"
          rows="10"
          value={this.state.info}
          onChange={this.onInfoChange}
        />
        <div>
          <button className="button">Save Project</button>
        </div>
      </form>
    );
  }
}

export default ExpenseForm;
