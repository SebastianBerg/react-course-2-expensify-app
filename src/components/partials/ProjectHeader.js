import React, { Component } from "react";
import { connect } from "react-redux";
import database from "../../firebase/firebase";
import { setCurrentProject } from "../../actions/projects";
import { startSetExpenses } from "../../actions/expenses";

export class ProjectHeader extends Component {
  constructor(props) {
    super(props);
    this.state = {
      projects: this.getProjects(this.props.auth.uid),
      project: { isLoading: true }
    };
  }

  calculateCurrentFunds() {}

  createSelectItems() {
    let projectArray = Array.from(this.state.projects);
    let items = [];
    if (projectArray.length !== 0) {
      for (let i = 0; i <= this.state.projects.length - 1; i++) {
        items.push(
          <option
            key={[i]}
            value={projectArray[i].name}
            id={projectArray[i].id}
          >
            {projectArray[i].name}
          </option>
        );
        //here I will be creating my options dynamically based on
        //what props are currently passed to the parent component
      }
      return items;
    }
  }

  handleProjectSelect = event => {
    const projectArray = Array.from(this.state.projects);
    const selectedIndex = event.target.options.selectedIndex;
    const projectId = event.target.options[selectedIndex].getAttribute("id");
    const projectIndex = this.state.projects.findIndex(x => x.id === projectId);
    const project = projectArray[projectIndex];
    this.props.setCurrentProject(project);
    this.props.startSetExpenses();
    console.log(this.props.history);
  };

  getProjects = uid =>
    database
      .ref(`users/${uid}/projects`)
      .once("value")
      .then(snapshot => {
        const projects = [];

        snapshot.forEach(childSnapshot => {
          projects.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        this.setState({ projects });
        this.setState({ project: { isLoading: false } });
      });

  componentWillMount() {
    this.getProjects(this.props.auth.uid);
  }

  render() {
    return (
      <div className="page-header">
        <div className="content-container">
          <div className="input-group__item header__content">
            <h1 className="page-header__title">
              {this.props.project.name ? (
                <span>{this.props.project.name}</span>
              ) : (
                <span>Select a Project</span>
              )}
            </h1>
            {!this.state.project.isLoading ? (
              <select
                className="select"
                onChange={this.handleProjectSelect}
                defaultValue={this.props.project.name || "default"}
              >
                <option disabled value="default">
                  Select a Project
                </option>
                {this.createSelectItems()}
              </select>
            ) : null}
          </div>

          <div className="header__content">
            <h1 className="page-header__title">
              Starting Funds: <span>{this.props.project.initialFunds}</span>
            </h1>
            <h1 className="page-header__title">
              Current Funds: <span>{this.props.project.initialFunds}</span>
            </h1>
          </div>
          {this.props.project.info ? (
            <div className="container-content">
              <h1 className="page-header__title--center">Additional Info:</h1>

              {this.props.project.info}
            </div>
          ) : null}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  project: state.project,
  auth: state.auth
});

const mapDispatchToProps = dispatch => ({
  setCurrentProject: project => dispatch(setCurrentProject(project)),
  startSetExpenses: () => dispatch(startSetExpenses())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProjectHeader);
