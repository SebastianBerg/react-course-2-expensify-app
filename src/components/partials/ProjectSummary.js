import React from "react";
import { connect } from "react-redux";

export class ProjectSummary extends React.Component {
  getTotalCost() {
    const projectArray = Array.from(this.props.expenses);
    if (projectArray) {
      const projectTotal =
        projectArray
          .map(expense => expense.amount)
          .reduce((prev, next) => prev + next) / 100;
      return projectTotal;
    }
  }

  getNumberExpenses() {
    const projectArray = Array.from(this.props.expenses);
    if (projectArray) {
      const expensesNumber = projectArray.length;
      return expensesNumber;
    }
  }

  render() {
    return (
      <div className="content-container">
        {this.props.expenses.length > 0 ? (
          <div>
            <h1 className="page-header__title--center">Project Summary</h1>
            <div className="header__content--s-padding">
              <h1 className="page-header__title">
                Number of expenses: <span>${this.getNumberExpenses()}</span>
              </h1>
              <h1 className="page-header__title">
                Total Cost: <span>${this.getTotalCost()}</span>
              </h1>
            </div>
          </div>
        ) : null}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  expenses: state.expenses,
  project: state.project
});

export default connect(mapStateToProps)(ProjectSummary);
