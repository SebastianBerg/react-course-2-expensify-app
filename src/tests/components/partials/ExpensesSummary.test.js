import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../../components/partials/ExpensesSummary";

test("should correctly render ExpensesSummary with 1 expense", () => {
  const totalExpense = 200.23;
  const wrapper = shallow(
    <ExpensesSummary expensesCount={1} expensesTotal={totalExpense} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should correctly render ExpensesSummary with multiple expenses", () => {
  const totalExpense = 20064646164;
  const wrapper = shallow(
    <ExpensesSummary expensesCount={23} expensesTotal={totalExpense} />
  );
  expect(wrapper).toMatchSnapshot();
});
