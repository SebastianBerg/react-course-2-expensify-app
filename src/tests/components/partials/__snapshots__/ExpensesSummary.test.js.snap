// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
     
    expenses
     totalling
     
    $2.00
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
     
    expenses
     totalling
     
    $200,646,461.64
  </h1>
</div>
`;
